File: mojo/edk/js/support.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/edk/js/support.h"

#include "base/bind.h"
#include "gin/arguments.h"
#include "gin/converter.h"
#include "gin/function_template.h"
#include "gin/object_template_builder.h"
#include "gin/per_isolate_data.h"
#include "gin/public/wrapper_info.h"
#include "gin/wrappable.h"
#include "mojo/edk/js/handle.h"
#include "mojo/edk/js/waiting_callback.h"
#include "mojo/public/cpp/system/core.h"

namespace mojo {
namespace js {

namespace {

WaitingCallback* AsyncWait(const gin::Arguments& args,
                           gin::Handle<HandleWrapper> handle,
                           MojoHandleSignals signals,
                           v8::Handle<v8::Function> callback) {
  return WaitingCallback::Create(args.isolate(), callback, handle, signals)
             .get();
}

void CancelWait(WaitingCallback* waiting_callback) {
  waiting_callback->Cancel();
}

gin::WrapperInfo g_wrapper_info = { gin::kEmbedderNativeGin };

}  // namespace

const char Support::kModuleName[] = "mojo/public/js/support";

v8::Local<v8::Value> Support::GetModule(v8::Isolate* isolate) {
  gin::PerIsolateData* data = gin::PerIsolateData::From(isolate);
  v8::Local<v8::ObjectTemplate> templ = data->GetObjectTemplate(
      &g_wrapper_info);

  if (templ.IsEmpty()) {
    templ = gin::ObjectTemplateBuilder(isolate)
                .SetMethod("asyncWait", AsyncWait)
                .SetMethod("cancelWait", CancelWait)
                .Build();

    data->SetObjectTemplate(&g_wrapper_info, templ);
  }

  return templ->NewInstance();
}

}  // namespace js
}  // namespace mojo

File: mojo/public/js/support.js
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// Module "mojo/public/js/support"
//
// Note: This file is for documentation purposes only. The code here is not
// actually executed. The real module is implemented natively in Mojo.

while (1);

/*
 * Waits on the given handle until the state indicated by |signals| is
 * satisfied.
 *
 * @param {MojoHandle} handle The handle to wait on.
 * @param {MojoHandleSignals} signals Specifies the condition to wait for.
 * @param {function (mojoResult)} callback Called with the result the wait is
 * complete. See MojoWait for possible result codes.
 *
 * @return {MojoWaitId} A waitId that can be passed to cancelWait to cancel the
 * wait.
 */
function asyncWait(handle, signals, callback) { [native code] }

/*
 * Cancels the asyncWait operation specified by the given |waitId|.
 * @param {MojoWaitId} waitId The waitId returned by asyncWait.
 */
function cancelWait(waitId) { [native code] }
